common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/MergedShuffleFileManager.java

```diff
@@ -0,0 +1,116 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.network.shuffle;
+
+import java.io.IOException;
+
+import org.apache.spark.network.buffer.ManagedBuffer;
+import org.apache.spark.network.client.StreamCallbackWithID;
+import org.apache.spark.network.shuffle.protocol.FinalizeShuffleMerge;
+import org.apache.spark.network.shuffle.protocol.MergeStatuses;
+import org.apache.spark.network.shuffle.protocol.PushBlockStream;
+
+
+/**
+ * The MergedShuffleFileManager is used to process push based shuffle when enabled. It works
+ * along side {@link ExternalBlockHandler} and serves as an RPCHandler for
+ * {@link org.apache.spark.network.server.RpcHandler#receiveStream}, where it processes the
+ * remotely pushed streams of shuffle blocks to merge them into merged shuffle files. Right
+ * now, push based shuffle can only be enabled when external shuffle service in YARN mode
```

Comment:
why is this only yarn?  what if I wrote one using external shuffle on standalone for instance, is there anything preventing me? Would be nice add to the comment as to why

Reply:
That comment might not be clear enough.
It's rather that right now we only implemented support for push-based shuffle when external shuffle service in YARN mode is used.
Extending to standalone mode is definitely doable, considering the similarity of the usage of ESS in that mode compared with YARN.
For K8S, we discussed briefly with @dbtsai and @holdenk  before.
They prefer to not use external shuffle service in K8S due to the limitation of mounting persistent volume in that environment, i.e. the volume can only be mounted in one pod for I/O.
So some work might be needed there to extend support into K8S.
There's also the need of supporting disaggregated shuffle storage, which we are also evaluating.
I have created SPARK-32925 for discussions on this.